### ISSUE TITLE
Support --mount,type=bind

### ISSUE THREAD

**Issue by jimmyharris** [earthly/earthly#1139]
In order to make use of large datasets that don't easily fit into docker containers at build time (and to reference devices in the host system at build time. We need a mechanism to support bind-mounts during run commands. This would allow us to access host filesystem resources that are more efficient than trying to import all of them via a COPY command.

**Comment by vladaionescu** [comment]
The use-case mentioned here is basically a faster `COPY`.

Over Slack you were mentioning access to `/dev/dri` as a need for bind mounts, right?

**Comment by jimmyharris** [comment]
Yes. alternate use case is to mount host system device files into the build image.

**Comment by jclangst** [comment]
Another important use case is that this would enable you more flexibility than the cache mount, especially for use cases where the cache file system is well structured and ideally can be shared across a wide array of builds (e.g., `nix` package manager, etc.)

**Comment by simonvanderveldt** [comment]
> The use-case mentioned here is basically a faster `COPY`.

I think in essence this is correct, but there's also some mental overhead/noise.
You have to include something like `COPY . .` if you want the equivalent of `docker run -v $PWD:<location>`. Which is just noise for the actual thing you want to do.
And this is ignoring the `SAVE ARTIFACT AS LOCAL` thing you'll also need to do if you expect some output.

**Comment by vladaionescu** [comment]
`bind-experimental` is not meant for regular use. It should have been probably called `bind-internal` and maybe not be exposed in the syntax at all :-) . And yeah, what it does is bind-mount stuff from the earthly-buildkit container.

**Comment by vladaionescu** [comment]
We've put together a proposal on a possible way to address the need for a local bind over here: https://github.com/earthly/earthly/issues/1399. Please leave feedback and/or 👍  over there if it serves your needs. Also, I'm curious about specific use-cases where it might not and local dir bind is strictly needed.

**Comment by BenAlanSouthall** [comment]
My usecase for wanting volume support is to enable easily rebuilding code when working on development. This also would probably link in with #1052. The `COPY` is slow for us, but even with caches properly set up, etc., I think our developers would appreciate the ability to drop into a container using the `-i` option, and just re-run make as many times as needed, inspect and mess around, until the code is ready to commit, in which case, running Earthly again but doing things "properly" using the `COPY` command would make sense. Perhaps this isn't the right way to do things, I'd be open to alternatives.

Personally, I'd probably prefer `VOLUME` to be expanded over the docker version (which is a bit useless, since it doesn't do much in the target image) to be a bit more like the new `CACHE` instruction. I.e. we can specify a volume for an entire stage, rather than always needing to append it to every `RUN` command, as it gets quite messy if you need it for more than one command.

**Comment by AlexMikhalev** [comment]
@NanoBjorn  @BenAlanSouthall You can create a cache for folders inside your build target:
```
RUN --mount=type=cache,target=/fluvio-duck/build/release/CMakeFiles/ /usr/bin/bash build_extension.sh
```
where `/fluvio-duck/build/release/CMakeFiles/` is cached. 
[Source](https://github.com/infinyon/fluvio-duck/blob/8f8a7648aa1cd60c318c04713ba348c1b5bc279b/Earthfile) is mix of CPP + Rust. Follow best practices for Earthly with +debs and caching  - I cut release compile time from 50 minutes to 25, so go [example](https://github.com/earthly/earthly/blob/main/examples/tutorial/go/part6/Earthfile) shall be fairly easy to extend.

**Comment by AlexMikhalev** [comment]
@NanoBjorn  @BenAlanSouthall You can create a cache for folders inside your build target:
```
RUN --mount=type=cache,target=/fluvio-duck/build/release/CMakeFiles/ /usr/bin/bash build_extension.sh
```
where `/fluvio-duck/build/release/CMakeFiles/` is cached. 
[Source](https://github.com/infinyon/fluvio-duck/blob/8f8a7648aa1cd60c318c04713ba348c1b5bc279b/Earthfile) is mix of CPP + Rust. Follow best practices for Earthly with +debs and caching  - I cut release compile time from 50 minutes to 25, so go [example](https://github.com/earthly/earthly/blob/main/examples/tutorial/go/part6/Earthfile) shall be fairly easy to extend.

**Comment by AlexMikhalev** [comment]
@NanoBjorn  @BenAlanSouthall You can create a cache for folders inside your build target:
```
RUN --mount=type=cache,target=/fluvio-duck/build/release/CMakeFiles/ /usr/bin/bash build_extension.sh
```
where `/fluvio-duck/build/release/CMakeFiles/` is cached. 
[Source](https://github.com/infinyon/fluvio-duck/blob/8f8a7648aa1cd60c318c04713ba348c1b5bc279b/Earthfile) is mix of CPP + Rust. Follow best practices for Earthly with +debs and caching  - I cut release compile time from 50 minutes to 25, so go [example](https://github.com/earthly/earthly/blob/main/examples/tutorial/go/part6/Earthfile) shall be fairly easy to extend.

**Comment by AlexMikhalev** [comment]
My apologies for the multiple comments - something went wrong between the browser and GitHub.

**Comment by Itxaka** [comment]
A couple of good use cases:

 - you use qemu to run your test isolated and to be repeatable, no matter the system is run from. No kvm available because there is no /dev/kvm
 - you run anything that needs access to a tpm device, like generating PCR values when building a UKI file. Impossible to do as there is no access to that device anywhere.


I think a way to expose certains mounts per target (or even globablly) would prove really useful.

**Comment by VasiliyS** [comment]
Another use case is having access to `.git` to e.g. generate dynamic version information.  Currently I do it with a `LOCALLY` target  but this has it's own limitations.
